### ISSUE TITLE
How do I use this?

### ISSUE THREAD

**Issue by konsumer** [MikeMnD/tiled-to-godot-export#5]
Here is what I tried:

* I have a simple map that uses terrain brushes
* I export the tileset with `projectRoot` set to `assets`
* I export map with `projectRoot` set to `assets`
* When I import scn file into godot, it says the asset can't be found, looks like it added the full path, so I set it to the project-relative path with "Fix Dependencies"
* I get another error the png file is missing (seems like it's looking for full-pth in `res://` again) so I hand-edit to 2 files to make sure the paths are correct

I still get this error:

```
 scene/resources/resource_format_text.cpp:175 - Couldn't load external resource: res://Terrain.tres
 editor/editor_data.cpp:798 - Index p_idx = 2 is out of bounds (edited_scene.size() = 2).
 editor/editor_data.cpp:627 - Index p_idx = 2 is out of bounds (edited_scene.size() = 2).
 editor/editor_data.cpp:677 - Index p_idx = 2 is out of bounds (edited_scene.size() = 2).
```

Is it because I'm using a map with terrain-brushes?

OSX: 10.14.6 (Mojave)
Tiled: 1.3.3
Godot: 3.2.1

Here is a test project:
[tester.zip](https://github.com/MikeMnD/tiled-to-godot-export/files/4473466/tester.zip)

**Comment by konsumer** [comment]
I noticed it also had the `0/region = Rect2( 0, 0, undefined, undefined )` problem, so I updated the tileset:

```
0/region = Rect2( 0, 0, 1024, 1024 )
```

I still had some of the issues, but tried with a map that only has 1 layer and no terrain brushes. I got further, but it still looks messed up:

<img width="1145" alt="Screen Shot 2020-04-13 at 11 47 10 PM" src="https://user-images.githubusercontent.com/83857/79194355-42cbfd80-7de1-11ea-9c43-50eb3ed403ae.png">

Here is the example project:
[Tiled Test.zip](https://github.com/MikeMnD/tiled-to-godot-export/files/4473593/Tiled.Test.zip)

**Comment by MikeMnD** [comment]
![image](https://user-images.githubusercontent.com/186191/79231178-583b3900-7e6e-11ea-89fc-8df9ca827439.png)

Hi just downloaded your Tiled Test.zip sample.
The only error I found is the project root should be an absolute path.
As you can see in my screenshot.
Then I just reexported your maps and all seems to work fine.

About the first sample with the terrain brushes - I'll have to look more since I don't even know what terrain brushes are :) so the export plugin doesn't support this for now.

**Comment by konsumer** [comment]
> The only error I found is the project root should be an absolute path.

I will do this from now on. Happy to add a PR for README.

> About the first sample with the terrain brushes - I'll have to look more since I don't even know what terrain brushes are :) so the export plugin doesn't support this for now.

No prob! Terrain brushes are cool.  If you look at the terrain sample in Tiled, you can choose "Terrain" in the tiles section, and it automatically figures out the correct shape and pulls the right tiles. It also has random variations it adds, if you set it up, like "Full Dirt" will randomly have little sprigs of grass, or it will randomly add other variations. It makes creating large areas less repetitive. There are some caveates, for sure, but it's cool if you set it up right. It's the main reason I want to use tiled over godot's built-in tile-editor (and also the slow tedious step of setting all the tiles in a sprite-image.)

[Here](https://doc.mapeditor.org/en/stable/manual/using-the-terrain-tool/) is more info about using them.

The main trick I have found is to create an "around a hole" set of 4, and a "inside hole" set of 4, and I like to use transparency and put each on it's own tile-layer, and then add "accent squares" for fill areas, so it works correctly:

![Screenshot from 2020-04-14 11-31-56](https://user-images.githubusercontent.com/83857/79260915-e77f2700-7e43-11ea-97c3-f222c6108cd8.png)

The ground here was done on a few tile-layers, and the dirt automatically added sprigs of grass, and the grass patches automatically drew the borders:

![Screenshot from 2020-04-14 11-33-27](https://user-images.githubusercontent.com/83857/79261037-172e2f00-7e44-11ea-9026-727678739456.png)

**Comment by konsumer** [comment]
Here is what that looks like in my editor:

![Screenshot from 2020-04-14 11-36-08](https://user-images.githubusercontent.com/83857/79261248-670cf600-7e44-11ea-9255-68daac012d0a.png)

**Comment by konsumer** [comment]
There is another technique where you can create a kind of auto-mapping atlas that lets you do it all on 1 layer, and it joins the transitions, (you have to pre-generate all the permutations) but I found it to be a bit buggy, so I just use 1 layer per terrain-type, with transparencies.

**Comment by konsumer** [comment]
Here is me adding a black hole (it handles all the borders, automatically):

![Peek 2020-04-14 11-40](https://user-images.githubusercontent.com/83857/79261726-1ba71780-7e45-11ea-95f2-b8c0902bf93d.gif)

**Comment by MikeMnD** [comment]
I've just tested the first sample and again fixing the projectRoot seems to fix the problem.

![image](https://user-images.githubusercontent.com/186191/79262313-c9441600-7e99-11ea-9fa6-a67da7085f54.png)

![image](https://user-images.githubusercontent.com/186191/79262283-be898100-7e99-11ea-93d0-b8ed2e99aa96.png)

Still, I don't understand how exactly the terrain tool works :P but gonna check it in the Tiled docs.

**Comment by konsumer** [comment]
You can also use Ctrl-key to make smaller "just borders" shapes:

![Peek 2020-04-14 11-47](https://user-images.githubusercontent.com/83857/79262371-0c749980-7e46-11ea-8899-915d6a12489f.gif)

**Comment by konsumer** [comment]
> I've just tested the first sample and again fixing the projectRoot seems to fix the problem.

awesome!

> Still, I don't understand how exactly the terrain tool works :P but gonna check it in the Tiled docs.

I think the output ends up being the same in the map, it just adds extra meta-info to the TIled tile definition (for editing.)

**Comment by MikeMnD** [comment]
Ah cool, so it's just nice tool which doesn't affect the export in any way :)

**Comment by konsumer** [comment]
I was about to PR the README, but it looks like you already added note for `projectRoot`. Awesome! I wonder if it might make sense to guess the projectRoot, based on the location of the tile-image vs the tile-definition & map vs tile-definition, as a default.  That way unmodified maps (no custom attribute) would work.

**Comment by konsumer** [comment]
It would also allow relative locations, which is good for putting this stuff in github (the user won;t have the same file-structure as me.)

**Comment by konsumer** [comment]
Maybe just a default of "`projectRoot` is the location of this file", so you can put them all in 1 dir, and it will figure it out. I think that is what tiled does (I like to put Tiled source in github projects for their games, and it works ok.) It would mean I have some extra tiled files in the godot project's dir, but I can live with that, and for compilation they could just be deleted. Should I PR for this?

**Comment by MikeMnD** [comment]
Well, I tried to get this but it's more of a Godot thing.
The project root is where your Godot project is Tiled doesn't know anything about that.
Maybe later on when projects are ready in Tiled we can move it as Project setting and there will be no need to be set on each Tilemap or Tileset individually.

**Comment by konsumer** [comment]
Yeh, this is totally workable as it is. Very easy to set 1 property to my maps/tile-sets. I wonder though if a sensible default like `res://` might be all it needs for it to work on github, with relative paths.

Maybe this logic would do it?

* check for `projectRoot`
* if it starts with `res://` just use that path directly in godot output (this would allow for godot subdirs and stuff)
* Use it as absolute, if it doesn't start with `.`
* If `projectRoot` is not set, set it to `res://` (this would allow tiled files in the godot project root)

This would handle these use-cases:

* I put all my files in godot's project-root (just leave `projectRoot` blank or set it to `res://`
* I put all my tiled files somewhere else, completely out of root (set `projectRoot` to an absolute path)
* I put all my files in some relative dir, that is maybe outside of godot's root (set  `projectRoot` to something like `../godot-game`
* I put all my files in some sub-dir of godot root (set `projectRoot` to `res://assets` or whatever)

**Comment by konsumer** [comment]
Just so we are speaking the same language with file-paths and stuff, I made a demo project [here](https://github.com/notnullgames/tiled-godot)

I still had issues with imported file. Is it due to multiple layers?

![Screenshot from 2020-04-14 13-13-32](https://user-images.githubusercontent.com/83857/79269952-1c927600-7e52-11ea-80dc-1832f57c4bd8.png)

**Comment by MikeMnD** [comment]
The link above doesn't work.
Multiple layers should not be a problem.
Can you show me the projectRoot you've set both for tileset and tilemap?

**Comment by konsumer** [comment]
> The link above doesn't work.

It works for me. It's on github:
https://github.com/notnullgames/tiled-godot

I even tried simplifying the map to just use 1 tileset (still with multiple layers) and it failed in the same way (weird lines.)

**Comment by MikeMnD** [comment]
Hm is it private repo?

**Comment by konsumer** [comment]
> Hm is it private repo?

Huh, yeh, it was. I'm a free-user, @notnullgames is my "game making org" so that's weird that it set it to private by default. Must be a new github thing. I made it public.

I just pushed a simplified map with just 1 tileset (but multiple layers)

Here is what the setting looks like in Tiled:

![Screenshot from 2020-04-14 13-27-06](https://user-images.githubusercontent.com/83857/79271389-9592cd00-7e54-11ea-890b-5725e45b856a.png)
![Screenshot from 2020-04-14 13-27-24](https://user-images.githubusercontent.com/83857/79271390-962b6380-7e54-11ea-8595-37e66d570bd4.png)

Of course, you will need to change it to your path make it work.

**Comment by konsumer** [comment]
It seems to be loading the tiles & layer-names correctly, just the mapping layer looks wrong:
![Screenshot from 2020-04-14 13-34-28](https://user-images.githubusercontent.com/83857/79271612-f02c2900-7e54-11ea-9560-c5d09343b4d6.png)

**Comment by MikeMnD** [comment]
Let me check the tilemap.tscn file that is generated

**Comment by MikeMnD** [comment]
Very strange I just downloaded change the two paths and exported.
Then opened Godot and all looks fine:
![image](https://user-images.githubusercontent.com/186191/79272099-986bdd00-7ea9-11ea-879c-05599b2370c0.png)

**Comment by MikeMnD** [comment]
The only difference I see is some blank lines in the .tscn file
Which may be some OS problem with the new lines...

**Comment by MikeMnD** [comment]
Ahhh there are even more differences:

[tiled-godot-master-working.zip](https://github.com/MikeMnD/tiled-to-godot-export/files/4477693/tiled-godot-master-working.zip)

**Comment by konsumer** [comment]
I made seperate issues so you can close this one.

* fix newlines on linux/mac #6
* `projectRoot` resolution: #7
